### ISSUE TITLE
[css-sizing] Last remembered size when fragmented?

### ISSUE THREAD

**Issue by Loirooriol** [w3c/csswg-drafts#7598]
https://drafts.csswg.org/css-sizing-4/#last-remembered

> record the current inner dimensions of its [principal box](https://drafts.csswg.org/css-display-3/#principal-box) as its [last remembered size](https://drafts.csswg.org/css-sizing-4/#last-remembered).

What if the principal box is fragmented e.g. into different columns when inside a multicol container?

ResizeObserver only reports the size in the 1st column: https://drafts.csswg.org/resize-observer-1/#resize-observer-entry-interface

> the current definitions of [content rect](https://drafts.csswg.org/resize-observer-1/#content-rect) and [border box](https://drafts.csswg.org/css-box-4/#border-box) do not mention how those boxes are affected by multi-column layout. In this spec, there will only be a single ResizeObserverSize returned in the [FrozenArray](https://webidl.spec.whatwg.org/#idl-frozen-array), which will correspond to the dimensions of the first column. A future version of this spec will extend the returned [FrozenArray](https://webidl.spec.whatwg.org/#idl-frozen-array) to contain the per-fragment size information.

So that's what Blink stores as the last remembered size.

Note that the last rememebred size is only used when the element has size containment, which makes it monolithic. So storing the sizes per fragment doesn't really make sense.

Just storing the size of the 1st fragment may be reasonable enough, but this should be clarified.

**Comment by Loirooriol** [comment]
I was actually mistaken. It's true that in Blink a ResizeObserver exposes an array with a single size, which is then used as the last remembered size. But this size is not the one of the 1st fragment as the spec says, in the block axis it's the sum of the block size of all fragments. So even if by accident, the last remembered size behaves as desired in Blink.

In Gecko I have implemented multi-fragment support in ResizeObserver (behind a flag), and then the last remembered size can use that.

So now I think it's reasonable to resolve that the last remembered size should take all fragments into account:
 - The last remembered block size should be the sum of the block sizes of the fragments.
 - The last remembered inline size should the the maximum (?) among the inline sizes of the fragments.

Not sure if Blink is actually using the maximum for the inline axis, since I don't know how to get fragments with different inline sizes.

**Comment by Loirooriol** [comment]
Test in http://wpt.live/css/css-sizing/contain-intrinsic-size/auto-010.html

**Comment by Loirooriol** [comment]
Regarding `box-decoration-break: clone`, Blink doesn't support it in the block axis. And as I said Gecko only uses the content size, so e.g. in a multicol which is 100px tall, an element with a 10px border and a content height of 125px will have 2 fragments: one with a content height of 80px / border height of 100px, and another with a content height of 45px / border height of 65px. Then the last remembered height will be 80px + 45px = 125px. Then when using the last remembered size, it will keep a content height of 125px.

In other words, when "stitching together in block axis", cloned box decorations are removed, then we take the bounding content size.

**Comment by cathiechen** [comment]
Should we split the content with last remembered size into different columns?

For instance, the first cast in
> Test in http://wpt.live/css/css-sizing/contain-intrinsic-size/auto-010.html

The last remembered size is 50 x 150, and the fragment container's height is 100px, the content could be split into two fragments 50 x 100 and 50 x 50, or like the expected 50 x 150.

I personally think two fragments makes more sense, the reason we use last remembered size is to reuse the old layout result. WDYT?

**Comment by Loirooriol** [comment]
I made the test like that because both Gecko and Blink do not fragment elements with `contain: size`: [testcase](https://software.hixie.ch/utilities/js/live-dom-viewer/?%3C!DOCTYPE%20html%3E%0A%3Cstyle%3E%0A%23wrapper%20%7B%0A%20%20column-width%3A%20100px%3B%0A%20%20width%3A%20max-content%3B%0A%20%20height%3A%20100px%3B%0A%7D%0A%23target%20%7B%0A%20%20contain%3A%20size%3B%0A%20%20width%3A%2050px%3B%0A%20%20height%3A%20150px%3B%0A%20%20background%3A%20orange%3B%0A%7D%0A%3C%2Fstyle%3E%0A%3Cdiv%20id%3D%22wrapper%22%3E%0A%20%20%3Cdiv%20id%3D%22target%22%3E%3C%2Fdiv%3E%0A%3C%2Fdiv%3E).

But from https://github.com/w3c/csswg-drafts/issues/7598#issuecomment-1240028670 it can be fragmented by slicing the contents, so WebKit is not wrong.

I guess the test can be relaxed to only check `clientHeight` and `clientWidth` when the element has size containment.

**Comment by cathiechen** [comment]
@Loirooriol Thanks for the explanation. Just created a PR for it. https://github.com/web-platform-tests/wpt/pull/37685
